process.c

```diff
@@ -618,18 +618,34 @@ proc_s_last_status(VALUE mod)
     return rb_last_status_get();
 }
 
-VALUE
-rb_process_status_new(rb_pid_t pid, int status, int error)
+static void
+process_status_initialize(struct rb_process_status *data, rb_pid_t pid, int status, int error)
 {
-    VALUE last_status = rb_process_status_allocate(rb_cProcessStatus);
-
-    struct rb_process_status *data = RTYPEDDATA_DATA(last_status);
     data->pid = pid;
     data->status = status;
     data->error = error;
+}
+
+static VALUE
+rb_process_status_initialize(VALUE self, VALUE pid, VALUE status, VALUE error)
+{
+    struct rb_process_status *data = RTYPEDDATA_DATA(self);
+
+    process_status_initialize(data, NUM2PIDT(pid), NUM2INT(status), NUM2INT(error));
 
-    rb_obj_freeze(last_status);
-    return last_status;
+    return self;
+}
+
+VALUE
+rb_process_status_new(rb_pid_t pid, int status, int error)
+{
+    VALUE instance = rb_process_status_allocate(rb_cProcessStatus);
+
+    struct rb_process_status *data = RTYPEDDATA_DATA(instance);
+    process_status_initialize(data, pid, status, error);
+
+    rb_obj_freeze(instance);
```

Comment:
This needs to move in `rb_process_status_initialize` like @nobu did. No reason to keep the instance mutable once it has been initialize'd.